### PR TITLE
BuildGraph for GoogleNet, DenseNet, ResNet, Yolo. Add MatMul_Layer, SoftMax_Layer, Reshape_Layer, BatchNormalization_Layer. Modificates for FCLayer, Flatten, Pooling, Convolution. Add universal accuracy check

### PR DESCRIPTION
**Pull request by Semyon1104**
The accuracy of alexnet_mnist remained the same at 98.01%

At the time of PR, the inference speed for one image is as follows: Googlenet - 1.5 seconds, Densenet - 2.8 seconds, Yolo - 7.6 seconds, Resnet - 14.6 seconds (after modifying the parallelism in the convolution layer, parallelization is performed not by batch size, but by output). It should also be noted that if the laptop is not charging, the time for each network increases by an average of 2 times. 

I think I'll need to add parallelism for EWLayer, as it takes a long time to run in some places

The test dataset currently contains 50,000 images, and I think that before running accuracy, we should try to to optimize some parts further, both in the graph assembly and image preparation, and possibly within the layers. For example, there is no parallel implementation for ConvDepthVise and EWlayer, which immediately comes to mind. Alternatively, we can check the accuracy on smaller parts of the dataset.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -59,7 +59,7 @@ jobs:
       with:
        name: mnist-${{ matrix.build_type }}${{ matrix.stats && '-stats' || '' }}
        path: |
-        build/bin/ACC_MNIST*
+        build/bin/ACC*
```
Comment (allnes) [inline .github/workflows/ci.yml:62]: Avoid wildcard in artifact path (`build/bin/ACC*`). Export a single `BINARY_PATH` from build and use it explicitly to prevent chmod/run from matching multiple binaries.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -274,7 +274,7 @@ jobs:
 
     - name: Prepare environment
       run: |
-        chmod +x build/bin/ACC_MNIST*
+        chmod +x build/bin/ACC*
```
Comment (allnes) [inline .github/workflows/ci.yml:277]: Replace `chmod +x build/bin/ACC*` with `chmod +x "$BINARY_PATH"`.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -290,11 +290,11 @@ jobs:
        export LD_LIBRARY_PATH=$PWD/build/bin/opencv_libs:/usr/lib/x86_64-linux-gnu
        echo "LD_LIBRARY_PATH: $LD_LIBRARY_PATH"
     
-       LD_DEBUG=files ./build/bin/ACC_MNIST* 2> ld_debug.log
+       LD_DEBUG=files ./build/bin/ACC* --model alexnet_mnist 2> ld_debug.log
        echo "### Library loading debug ###"
        grep -i "opencv_imgcodecs" ld_debug.log
 
-       ./build/bin/ACC_MNIST* > accuracy.txt
+       ./build/bin/ACC* > accuracy.txt
```
Comment (allnes) [inline .github/workflows/ci.yml:297]: Avoid `./build/bin/ACC*` wildcard for accuracy run; use the explicit binary path

File: app/Converters/parser_onnx.py [hunk 4]
```diff
@@ -5,15 +5,17 @@
...
 from ultralytics import YOLO
 
+
 def convert_pt_to_onnx(pt_model_path, onnx_model_path=None):
     if onnx_model_path is None:
         onnx_model_path = pt_model_path.replace('.pt', '.onnx')
 
     model = YOLO(pt_model_path)
-    model.export(format="onnx", dynamic=False, simplify=True)
+    model.export(format="onnx", dynamic=False, simplify=False)
```
Comment (allnes) [inline app/Converters/parser_onnx.py:14]: Why `simplify=False`? If simplification caused issues, add a comment or a flag to control it; otherwise prefer `simplify=True` for a smaller, stable graph.

File: app/Converters/parser_onnx.py [hunk 5]
```diff
@@ -31,12 +33,29 @@ def onnx_to_json(model_path, output_json_path):
...
     layer_info = []
+
+    input_info = {}
+    for input in model.graph.input:
+        if input.name in initializers_dict:
+            continue
+
+        input_info = {
+            "name": input.name,
+            "shape": [dim.dim_value for dim in input.type.tensor_type.shape.dim],
```
Comment (allnes) [inline app/Converters/parser_onnx.py:44]: `dim.dim_value` is 0 for dynamic dims. Consider handling `dim_param` or mapping 0→-1/None to preserve dynamic shape semantics in JSON.

File: app/Converters/parser_onnx.py [hunk 6]
```diff
@@ -112,6 +151,7 @@ def default(self, obj):
         json.dump(layer_info, f, indent=2, cls=CustomEncoder)
 
     print(f"Модель успешно сохранена в {output_json_path}")
+    print(f"Input shape: {input_info.get('shape', [])}")
```
Comment (allnes) [inline app/Converters/parser_onnx.py:154]: Library code prints to stdout. Prefer returning metadata or using a logger to avoid noisy CLI output.

File: app/Graph/build.cpp [hunk 7]
```diff
@@ -200,4 +193,1114 @@ void build_graph(it_lab_ai::Tensor& input, it_lab_ai::Tensor& output,
       }
     }
   }
+}
+
+std::string get_base_layer_name(const std::string& tensor_name) {
```
Comment (allnes) [inline app/Graph/build.cpp:198]: `std::regex` compiled on every call. Make it `static const` or cache it.

File: app/Graph/build.cpp [hunk 8]
```diff
@@ -200,4 +193,1114 @@ void build_graph(it_lab_ai::Tensor& input, it_lab_ai::Tensor& output,
...
+
+  if (!file.is_open()) {
+    throw std::runtime_error("Cannot open class names file: " + filename);
+  }
+
+  while (std::getline(file, line)) {
+    line = std::regex_replace(line, std::regex("^\\s+|\\s+$"), "");
+    if (line.empty()) continue;
+
+    std::regex pattern("(\\d+):\\s*'([^']+)'");
```
Comment (allnes) [inline app/Graph/build.cpp:1149]: Compiling `std::regex` inside a per-line loop in `load_class_names`. Move the regex creation outside the loop.

File: include/layers/Tensor.hpp [hunk 9]
```diff
@@ -21,11 +21,12 @@ template <typename T>
 const std::vector<uint8_t>* to_byte(const std::vector<T>& v) {
   return reinterpret_cast<const std::vector<uint8_t>*>(&v);
 }
-
 template <typename T>
 Type GetTypeEnum() {
   if constexpr (std::is_same_v<T, int>) {
     return Type::kInt;
+  } else if constexpr (std::is_same_v<T, int64_t>) {
```
Comment (allnes) [inline include/layers/Tensor.hpp:28]: `GetTypeEnum<int64_t>()` returns `Type::kInt` (same as 32‑bit). This loses int32/int64 distinction; consider a separate enum value or explicit casts at boundaries.

File: include/layers/SoftmaxLayer.hpp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+#include <vector>
+
+#include "layers/Layer.hpp"
+#include "layers/Tensor.hpp"
+
+namespace it_lab_ai {
+
+class SoftmaxLayer : public Layer {
+ public:
+  explicit SoftmaxLayer(int axis = -1) : axis_(axis) {}
```
Comment (allnes) [inline include/layers/SoftmaxLayer.hpp:14]: Add tests for negative/large `axis`, axis normalization, and numerical stability (subtract max to prevent overflow).

File: include/layers/ReshapeLayer.hpp [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+#include <vector>
+
+#include "layers/Layer.hpp"
+#include "layers/Tensor.hpp"
+
+namespace it_lab_ai {
+
+class ReshapeLayer : public Layer {
+ public:
+  explicit ReshapeLayer(bool allowzero = false,
```
Comment (allnes) [inline include/layers/ReshapeLayer.hpp:11]: ONNX `allowzero`: when 1, zeros in target shape copy input dims. Ensure this semantics and validate product with `-1`; add tests.

File: include/layers/PoolingLayer.hpp [hunk 12]
```diff
@@ -117,66 +196,102 @@ std::vector<ValueType> PoolingLayerImpl<ValueType>::run(
...
+          if (batch_dim >= 0) output_coords[batch_dim] = n;
+          if (channel_dim >= 0) output_coords[channel_dim] = c;
+          output_coords[this->outputShape_.dims() - spatial_dims] = h;
+          if (spatial_dims > 1)
+            output_coords[this->outputShape_.dims() - spatial_dims + 1] = w;
+
+          size_t output_index = this->outputShape_.get_index(output_coords);
+
+          if (!pooling_buf.empty()) {
+            switch (this->poolingType_) {
```
Comment (allnes) [inline include/layers/PoolingLayer.hpp:268]: Add a `default:` in the `switch (poolingType_)` to guard unknown enums; same in the TBB variant.

File: app/Graph/build.cpp [hunk 13]
```diff
@@ -193,4 +186,1064 @@ void build_graph(it_lab_ai::Tensor& input, it_lab_ai::Tensor& output,
...
+    case it_lab_ai::kReduce:
+      return "Reduce";
+    case it_lab_ai::kBatchNormalization:
+      return "BatchNormalization";
+    default:
+      return "Unknown";
+  }
+}
+
+void build_graph(it_lab_ai::Tensor& input, it_lab_ai::Tensor& output,
```
Comment (NeiroYT) [inline app/Graph/build.cpp]: please consider shortening this function, for example giant `for (const auto& layer_data : model_data)` parsing can be moved away to different function or even class.
main idea of build_graph is not parsing after all

File: test/single_layer/test_flattenlayer.cpp [hunk 14]
```diff
@@ -5,81 +5,263 @@
...
   }
-  std::vector<float> expected_2 = {0.0f, 12.0f, 1.0f,  13.0f, 2.0f,  14.0f,
-                                   3.0f, 15.0f, 4.0f,  16.0f, 5.0f,  17.0f,
-                                   6.0f, 18.0f, 7.0f,  19.0f, 8.0f,  20.0f,
-                                   9.0f, 21.0f, 10.0f, 22.0f, 11.0f, 23.0f};
-  std::vector<float> expected_3 = {0.0f,  6.0f,  1.0f,  7.0f,  2.0f,  8.0f,
-                                   3.0f,  9.0f,  4.0f,  10.0f, 5.0f,  11.0f,
-                                   12.0f, 18.0f, 13.0f, 19.0f, 14.0f, 20.0f,
-                                   15.0f, 21.0f, 16.0f, 22.0f, 17.0f, 23.0f};
+
```
Comment (NeiroYT) [inline test/single_layer/test_flattenlayer.cpp]: why test of correctness of flatten layer was removed?

Comment (Semyon1104) [reply test/single_layer/test_flattenlayer.cpp]: fixed

File: app/Graph/build.cpp (hunk 13 above)
Comment (Semyon1104) [reply app/Graph/build.cpp]: fixed
